#160451607 Users should be able to receive notifications

Victor-Ugwueze opened the pull request:
#### What does this PR do?
- send email and in-app-notifications
   
#### Description of Task to be completed?
- write tests for endpoints
- create NotificationController to prepare and dispatch notifications
- refactor EmailController to be reusable across module
- add SettingsController to update user notification settings

- create endpoints
      ` GET /api/v1/me/notifications/:id`
      ` GET /api/v1/me/notifications`
      ` PUT /api/v1/me/notifications/:id`
      `PUT /api/v1/me/settings/notifications`

#### How should this be manually tested?
 - run the app
 - make a post request to the above endpoints
 - required params and response 

##### Request and expected responses

 - request 
  ```js
 GET /api/v1/me/notifications
```   
  - response   
 ```js
{
    "status": "Success",
    "notifications": {
        "count": 0,
        "rows": []
    }
}
```

 - Follow a user with the below endpoint 
  - request 
  ```js
 POST /api/v1/profiles/1/follow
```
 -response
```js
{
    "profile": {
        "username": "DevOps",
        "image": null,
        "bio": null,
        "following": true
    }
}
```
   
then make a request 
   - request 
  ```js
 GET /api/v1/me/notifications/:id
```   
   - response   
 ```js
{
    "status": "Success",
    "notifications": {
        "id": 4,
        "isRead": true,
        "message": "Hey Okafor, chisom\n              started following you,\n              ",
        "createdAt": "2018-10-25T11:04:28.459Z",
        "updatedAt": "2018-10-25T11:09:45.353Z",
        "userId": 1
    }
}
```

  - request 
  ```js
 PUT /api/v1/me/notifications/:id
```   
  - response   
 ```js
{
    "status": "Succes",
    "message": "Notification marked as read"
}
```


 - request 
  ```js
 PUT /api/v1/me/settings/notifications
{
notificationType: "follow",
subscribe: false,
}
```   
  - response   
 ```js
{
    "status": "Success",
    "message": "You have subscribed to receiving follow notifications"
}
```

#### Any background context you want to provide?
 NA

#### What are the relevant pivotal tracker stories?
[#160451607](https://www.pivotaltracker.com/story/show/160451607 "send notification")
 
### Question
 NA

### Screenshot
<img width="1307" alt="screen shot 2018-10-25 at 12 33 24 pm" src="https://user-images.githubusercontent.com/24890605/47497514-3dc46880-d852-11e8-8995-ab65e62bb834.png">

Victor-Ugwueze committed:
feat(email and in-app notification): users should be able to receive notifications

- add NotificationController
   * users should receive notification when an author they are
     following publish article
   * a user should should receive notification when he gets a new follower
   * users should receive notification when a comment or reaction is added to their favourite article

- add SettingsController
  * a user should be able to optout of a specific notification

- refactor EmailController
  * refctor sendMail to send sigle or multiple email

- fix test failure after rebasing

[Delivers #160451607]

Victor-Ugwueze committed:
feat(email and in-app notification): users should be able to receive notifications

 - remove all fixtures
 - add subscribe to notification settings

[Delivers #160451607]

KaiserPhemi reviewed (changes requested):
#### Critical
Please rebase

Victor-Ugwueze committed:
feat(email and in-app notification): users should be able to receive notifications

- fix merge conflicts
- change id to slug on favourite article params

Delivers #160451607]

Victor-Ugwueze commented:
@KaiserPhemi, I have implemented the feedback. Please review thank you.
